File: jiuzhang/56_twoSum.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;

class Solution {
public:
    /**
     * @param numbers: An array of Integer
     * @param target: target = numbers[index1] + numbers[index2]
     * @return: [index1, index2] (index1 < index2)
     */
    vector<int> twoSum(vector<int> &numbers, int target) {
        // write your code here
        unordered_map<int, int> hashTable;
        for (int i = 0; i < numbers.size(); i++) {
            auto it = hashTable.find(target - numbers[i]);
            if (it != hashTable.end()) {
                return {it->second, i};
            }
            hashTable[numbers[i]] = i;
        }
        return {};
    }
};

File: jiuzhang/56_twoSum.py
from typing import (
    List,
)
#lintcode 56 
class Solution:
    """
    @param numbers: An array of Integer
    @param target: target = numbers[index1] + numbers[index2]
    @return: [index1, index2] (index1 < index2)
    """
    def two_sum(self, numbers: List[int], target: int) -> List[int]:
        # write your code here
        res_dict = {}
        for i, num in enumerate(numbers):
            if target - num in res_dict:
                return [res_dict[target - num], i]
            res_dict[num] = i
        return []